chain/events/filter/index.go

```diff
@@ -56,27 +61,177 @@ var ddls = []string{
     	version UINT64 NOT NULL UNIQUE
 	)`,
 
-	// version 1.
 	`INSERT OR IGNORE INTO _meta (version) VALUES (1)`,
+	`INSERT OR IGNORE INTO _meta (version) VALUES (2)`,
 }
 
-const schemaVersion = 1
+var (
+	log = logging.Logger("filter")
+)
 
 const (
-	insertEvent = `INSERT OR IGNORE INTO event
-	(height, tipset_key, tipset_key_cid, emitter_addr, event_index, message_cid, message_index, reverted)
-	VALUES(?, ?, ?, ?, ?, ?, ?, ?)`
+	schemaVersion = 2
 
-	insertEntry = `INSERT OR IGNORE INTO event_entry
-	(event_id, indexed, flags, key, codec, value)
-	VALUES(?, ?, ?, ?, ?, ?)`
+	eventExists          = `SELECT MAX(id) FROM event WHERE height=? AND tipset_key=? AND tipset_key_cid=? AND emitter_addr=? AND event_index=? AND message_cid=? AND message_index=?`
+	insertEvent          = `INSERT OR IGNORE INTO event(height, tipset_key, tipset_key_cid, emitter_addr, event_index, message_cid, message_index, reverted) VALUES(?, ?, ?, ?, ?, ?, ?, ?)`
+	insertEntry          = `INSERT OR IGNORE INTO event_entry(event_id, indexed, flags, key, codec, value) VALUES(?, ?, ?, ?, ?, ?)`
+	revertEventsInTipset = `UPDATE event SET reverted=true WHERE height=? AND tipset_key=?`
+	restoreEvent         = `UPDATE event SET reverted=false WHERE height=? AND tipset_key=? AND tipset_key_cid=? AND emitter_addr=? AND event_index=? AND message_cid=? AND message_index=?`
 )
 
 type EventIndex struct {
 	db *sql.DB
+
+	stmtEventExists          *sql.Stmt
+	stmtInsertEvent          *sql.Stmt
+	stmtInsertEntry          *sql.Stmt
+	stmtRevertEventsInTipset *sql.Stmt
+	stmtRestoreEvent         *sql.Stmt
 }
 
-func NewEventIndex(path string) (*EventIndex, error) {
+func (ei *EventIndex) initStatements() (err error) {
+	ei.stmtEventExists, err = ei.db.Prepare(eventExists)
+	if err != nil {
+		return xerrors.Errorf("prepare stmtEventExists: %w", err)
+	}
+
+	ei.stmtInsertEvent, err = ei.db.Prepare(insertEvent)
+	if err != nil {
+		return xerrors.Errorf("prepare stmtInsertEvent: %w", err)
+	}
+
+	ei.stmtInsertEntry, err = ei.db.Prepare(insertEntry)
+	if err != nil {
+		return xerrors.Errorf("prepare stmtInsertEntry: %w", err)
+	}
+
+	ei.stmtRevertEventsInTipset, err = ei.db.Prepare(revertEventsInTipset)
+	if err != nil {
+		return xerrors.Errorf("prepare stmtRevertEventsInTipset: %w", err)
+	}
+
+	ei.stmtRestoreEvent, err = ei.db.Prepare(restoreEvent)
+	if err != nil {
+		return xerrors.Errorf("prepare stmtRestoreEvent: %w", err)
+	}
+
+	return nil
+}
+
+func (ei *EventIndex) migrateToVersion2(ctx context.Context, chainStore *store.ChainStore) error {
+	if chainStore == nil {
+		return errors.New("chainStore is nil")
+	}
```

Comment:
Unless we pass a nil chainstore on a happy path, I wouldn't bother with this check. From a code-style standpoint, I see this and start wondering if there might be reasonable cases where this is null.